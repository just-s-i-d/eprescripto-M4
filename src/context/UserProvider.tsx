import {
  PropsWithChildren,
  SetStateAction,
  createContext,
  useEffect,
  useState,
} from "react";

import { getUserData } from "@utils/Doctor";
import { showToast } from "@utils/common";

type UserContextType = {
  isLoggedIn?: boolean;
  setIsLoggedIn?: React.Dispatch<SetStateAction<boolean>>;
  isDarkTheme?: boolean;
  setIsDarkTheme?: React.Dispatch<SetStateAction<boolean>>;
  role?: string;
  setRole?: React.Dispatch<SetStateAction<string | undefined>>;
  refresh?: boolean;
  setRefresh?: React.Dispatch<SetStateAction<boolean | undefined>>;
  getRole?: () => void;
};

export const UserContext = createContext<UserContextType>({});
export const UserProvider = ({ children }: PropsWithChildren) => {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [isDarkTheme, setIsDarkTheme] = useState<boolean>(false);
  const [role, setRole] = useState<string>();

  const getRole = () => {
    getUserData()
      .then((res) => {
        setRole(res.role?.type);
      })
      .catch((error) => {
        showToast(error.message, error.type);
      });
  };
  useEffect(() => {
    const token = localStorage.getItem("token");
    if (token) {
      getRole();
    }
    setIsLoggedIn(!!token);
  }, []);
  const value = {
    isLoggedIn,
    setIsLoggedIn,
    role,
    setRole,
    isDarkTheme,
    setIsDarkTheme,
    getRole,
  };
  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};
