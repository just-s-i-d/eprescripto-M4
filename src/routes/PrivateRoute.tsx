import { useContext } from "react";
import { Navigate, Outlet } from "react-router-dom";

import { UserContext } from "@context/UserProvider";
import Loading from "@components/ui/Loading";

type PrivateRouteProps = {
  allowedRoles: string[];
};
const PrivateRoute = ({ allowedRoles }: PrivateRouteProps) => {
  const userContext = useContext(UserContext);

  if (!userContext) return <Loading />;
  const { isLoggedIn, role } = userContext;

  return (
    <>
      {!isLoggedIn && <Navigate to="/auth" />}
      {isLoggedIn && !allowedRoles.includes(role) && (
        <Navigate to="/forbidden" />
      )}
      {isLoggedIn && allowedRoles.includes(role) && <Outlet />}
    </>
  );
};

export default PrivateRoute;
