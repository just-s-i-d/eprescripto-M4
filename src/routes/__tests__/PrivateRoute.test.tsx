import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import PrivateRoute from "@routes/PrivateRoute";
import { UserContext } from "@context/UserProvider";

type WrapperComponentProps = {
  loggedIn: boolean;
  role: string;
};
jest.mock("@context/UserProvider");
const WrapperComponent = ({ loggedIn, role }: WrapperComponentProps) => {
  return (
    <MemoryRouter initialEntries={["/dashboard"]}>
      <UserContext.Provider value={{ isLoggedIn: loggedIn, role: role }}>
        <Routes>
          <Route
            element={<PrivateRoute allowedRoles={["doctor", "patient"]} />}
          >
            <Route path="dashboard" element={<h1>Dashboard</h1>} />
          </Route>
          <Route path="auth" element={<h1>Auth Page</h1>} />
          <Route path="forbidden" element={<h1>Forbidden Access Page</h1>} />
        </Routes>
      </UserContext.Provider>
    </MemoryRouter>
  );
};
describe("Private Route tests", () => {
  it("Render to dashboard when user is logged in and has access to dashboard", () => {
    const { asFragment } = render(
      <WrapperComponent loggedIn={true} role="doctor" />,
    );
    expect(screen.getByText("Dashboard")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Redirect to auth page test", async () => {
    render(<WrapperComponent loggedIn={false} role="doctor" />);
    expect(screen.getByText("Auth Page")).toBeInTheDocument();
  });
  it("Redirect to forbidden access page", async () => {
    render(<WrapperComponent loggedIn={true} role="notAllowedRole" />);
    expect(screen.getByText("Forbidden Access Page")).toBeInTheDocument();
  });
});
