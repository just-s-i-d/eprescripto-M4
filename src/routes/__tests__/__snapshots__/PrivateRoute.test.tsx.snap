// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Private Route tests Render to dashboard when user is logged in and has access to dashboard 1`] = `
<DocumentFragment>
  <h1>
    Dashboard
  </h1>
</DocumentFragment>
`;
